### ISSUE TITLE
DNSSEC Security Vulnerability NSEC

### ISSUE THREAD

**Issue by hsntgm** [webmin/webmin#511]
Hello @jcameron

Webmin defaultly setup DNSSEC with with NSEC signed.

This is a know thing which NSEC3 support more security over NSEC.

For NSEC enumeration, you can create the full list of domains by starting to guess at possible names in the domain. If the zone has around 100 domain names, it will take around 100 requests to enumerate the entire zone. With NSEC3, when you request a record that does not exist, a signed NSEC3 record is returned with the next zone present ordered alphabetically by hash. Checking if the next query name candidate fits in one of the known gaps allows anyone to discover the full chain in around 100 queries. 

**There are many tools that can do this computation for you, including a plug-in to nmap.** 

With the hashes that correspond to all the valid names in the zone, a dictionary attack can be used to figure out the real names. Short names are easily guessed, and by using a dictionary, longer names can be revealed as existing without having to flood the authoritative nameservers with guesses. Tools like HashCat..

In short, NSEC is like revealing plaintext passwords, and NSEC3 is like revealing a Unix-style passwords file.

You can find signing with NSEC3 intruction here.

https://deepthought.isc.org/article/AA-00711/0/In-line-Signing-With-NSEC3-in-BIND-9.9-A-Walk-through.html

Sincerely,

**Comment by jcameron** [comment]
Virtualmin doesn't actually control whether to create `NSEC` or `NSEC3` records - this is determined by the BIND tools on your system. Which linux distribution and version are  you runnijng there ... and have you upgraded all packages?

**Comment by hsntgm** [comment]
I am on ubuntu 16.04.2 server edition and i have NSEC records.All packets upgraded.

**Comment by jcameron** [comment]
When you created the zone signing key, which algorithm did you select?

**Comment by hsntgm** [comment]
May i miss anything ? I didn't choose any Algorithm while signing to zone. My DNSSEC setup automated on webmin framewok.

I used **Webmin > Bind Dns Server > DNSSEC Verification** to auto sign. After that i used **Setup DNSSEC Key** for my master zone this is also automated.  I didn't use any inline terminal code or select any Algorithm or any NSEC,NSEC3 options. So i declared it default webmin setup. I learnt that this operation managed by BIND tools not virtualmin by you. But how bind tools decide to sign it with nsec or nsec3 ?

If this help you my current setup has:
Algorithm: 5 (RSASHA1)
Created: 20170301

Best Regards,

**Comment by jcameron** [comment]
Did you create this zone using Webmin's BIND DNS Server module, or was it a Virtualmin domain?

In the BIND module, on the "Setup DNSSEC Key" page you can select a key algorithm.

**Comment by hsntgm** [comment]
It was a fresh webmin seup, first i created virtualmin domain sure, then I used Webmin > Bind Dns Server > DNSSEC Verification.There was a button something like "Webmin can automatically setup DLV for you" i used that.
I choosed below options.

`DNSSEC enabled? 	Yes 	DNSSEC response validation enabled? 	Yes`

After that i go to domain zone in same BIND Server Module and created key with Setup DNSSEC Key option. I didn't see any selectable a key algorithm. 

I am looking DNSSEC Key option area now and i see Remove Key ,Sign Zone, Re-Sign Zone options. 
You mean NSEC or NCES3 setup depends to algorithm selection in Setup DNSSEC Key?

Sincerely,

**Comment by hsntgm** [comment]
Ok, i tested it and it was my fault there is a options NSEC3DSA and NSEC3RSA algorithm options. This is deep in deep option to find it. 

Thank you for your forwarding.
Sincerely,

**Comment by iliaross** [comment]
❗️ If the issue isn't resolved, feel free to reopen it with updated details.
